### PR TITLE
[rush] Basic support of `--watch` for BulkScriptAction (build, rebuild, etc.)

### LINKED ISSUE #1122: [rush] make incremental build happen on file changes (watch mode)
I'd love a feature where you would type something like rush watch, and then it would invoke rush build whenever it sees that a file has been touched. This would be amazingly performant for a large repo.

Behavior would be is somewhat similar to how nodemon works. Some suggestions by current lerna maintainer here https://github.com/lerna/lerna/issues/1722#issuecomment-431069978

The main motivation behind this is that we'd be able to skip having watch instances running for each package, and instead have one watch at the top.

I think this would be a big step forward for scaled micro frontend monorepos. We have 30 packages now and the RAM usage of each is killing those poor developer laptops (for each package, npm is 20 mb, the process itself a minimum of 60mb).

Since rush already features incremental builds, and building upstream and downstream, I think this would be the perfect thing to build onto what is in Rush.

Full disclaimer, I'm not using rush yet but Lerna. This feature would tip the scale for me use rush, no doubt.

Another minor thing, I'd love to be able to have this watcher actually start watch instances of webpack. I know this will block execution, so some sort of daemonization would be necessary, and likely not something that rush could offer without some complementary implementation in each micro app. I thought perhaps it would be neat if rush could help the daemonization process by either documenting it, or letting the micro apps do process.send() with some daemon activation message. This area isn't my strong suit but this is a need that we also have, and that I plan to document here later on when it's done in my project (currently using lerna but that doesn't matter in this context).

### PR DESCRIPTION
**Pull request by dmichon-msft**
## Summary

Globally adds `--watch` support to BulkScriptAction, thereby including `rush build`, `rush rebuild`, and any user-defined commands that may do so more efficiently (for example someone may define `rush build:incremental`).

This allows any repository that defines `npm run build` to get the full benefit of automatic rebuild in response to source code changes, without regard to what tool is used for the individual project's build command.

Works best if the `npm run build` (or alternative command) script is optimized for efficient incremental rebuild with disk caching. Consider TypeScript 4, Webpack 5, etc.

Fixes #1122
Addresses "Feature 1" from #1202.

### Handling of `--watch`
Support for `--watch` is implemented in `BulkScriptAction` by evaluating the command line parameters to determine the selected projects, then invoking either a single execution or watch mode.
Watch mode is implemented as a basic loop:
1) Wait for one or more projects to have different inputs than the last recorded state, then record the state (as determined by `PackageChangeAnalyzer`). Initially there is no recorded state, so this returns immediately.
2) Run the underlying command, filtered to the changed projects and their dependents, filtered to the initial selection from the enclosing command
3) Goto (1)

### Change detection is implemented as follows:
1) Instantiate a new `PackageChangeAnalyzer` and compare the hashes for the selected projects with the previous (in-memory) execution state. If any projects contain changed hashes, skip directly to (4).
2) Instantiate a file system watcher using `chokidar` for the root folders of all selected projects. Any changes trigger a debounced check identical to (1), until such time as one or more projects report changes.
3) Shut down the file system watcher to save resources and avoid noise during build
4) Return the set of changed projects and the new state (`PackageChangeAnalyzer`) to hand to `TaskSelector`

### Backwards compatibility concerns:
- If, for some reason, developers were trying to forward `--watch` through a `BulkScriptAction` to underlying npm scripts, that will no longer work.

~~Currently includes the changes from #2422, will be rebased if that goes in. The `Selection` manipulation and refactoring of `TaskSelector` were relevant to implementation.~~

### Alternatives considered:
- *Communicate with individual package builder processes in watch mode.* This makes demands on how developers implement `npm run build`, and may produce faster results in certain optimal environments, but limits the general utility of the feature.
- *Use a separate action altogether, e.g. `rush build:watch`.* This makes parameter forwarding a bit awkward and would require either architectural changes in `command-line.json` for invoking an npm command with a different name than the rush command or for developers to supply an additional npm command definition to `build`. Not infeasible, but is a separate issue and alters the ergonomics of the command line. The current implementation conforms to the contract that `rush build --foo --bar` has the same behavior as `rush build --foo --bar --watch` with respect to building each project.

### Future improvements:
- Canceling extant build queue and immediately beginning a new build upon subsequent file changes to minimize total end to end time. This involves passing a cancellation token of some sort to `TaskRunner`, which currently has no precedence in the `rushstack` repo. Also requires the file system watcher to continue running during the build process, though this has not been shown to cause any particular problems.
- Phased and/or sharded builds are entirely within the domain of the `TaskSelector` and `TaskRunner` and should require no additional work with this implementation

## How it was tested

Invoked `rush build --to @microsoft/rush-lib --watch` against rushstack repo and observed rebuild of the changed project and downstream projects when changes were saved to disk.
Performing further file system changes during execution results in execution occurring immediately after the previous build pass completes.

Edit: Now that #2422 is in, contains only the directly relevant changes.

### CODE REVIEW

File: apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts [hunk 1]
```diff
@@ -329,6 +392,13 @@ export class BulkScriptAction extends BaseScriptAction {
         ' For details, refer to the website article "Selecting subsets of projects".'
     });
 
+    this._watchParameter = this.defineFlagParameter({
+      parameterLongName: '--watch',
+      parameterShortName: '-w',
+      description:
+        'Activates a file system watcher to dynamically re-invoke the command when projects change.'
```
Comment (apostolisms) [inline apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts:399]: Consider adding more detail/making it more clear as to what it is that --watch is doing? 
Watches for file changes and dynamically re-invokes the command that --watch is used with. 
Don't like this that much either. Maybe we only describe it in terms of rush build to provide clarity?

File: apps/rush-lib/src/cli/test/__snapshots__/CommandLineHelp.test.ts.snap [hunk 2]
```diff
@@ -445,6 +447,8 @@ Optional arguments:
                         each of the projects belonging to VERSION_POLICY_NAME.
                          For details, refer to the website article \\"Selecting 
                         subsets of projects\\".
+  -w, --watch           Activates a file system watcher to dynamically
```
Comment (apostolisms) [inline apps/rush-lib/src/cli/test/__snapshots__/CommandLineHelp.test.ts.snap:450]: same as previous comment in terms of description

File: apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts (hunk 1 above)
Comment (dmichon-msft) [reply apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts]: Updated a bit. We will need to add a companion article.

**Review by apostolisms — APPROVED** [review]

**Comment by octogonz** [comment]
>     1. Wait for one or more projects to have different inputs than the last recorded state, then record the state (as determined by `PackageChangeAnalyzer`). Initially there is no recorded state, so this returns immediately.

@dmichon-msft As I recall `PackageChangeAnalyzer` has trouble with these cases:
- A file that is tracked by Git, but which has been deleted locally (and not committed yet)
- A newly created file that has not been added to Git yet

Might be useful to test these edge cases, and if they don't work, we can document that.

File: apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts [hunk 3]
```diff
@@ -329,6 +392,18 @@ export class BulkScriptAction extends BaseScriptAction {
         ' For details, refer to the website article "Selecting subsets of projects".'
     });
 
+    this._watchParameter = this.defineFlagParameter({
+      parameterLongName: '--watch',
+      parameterShortName: '-w',
+      description:
+        'Normally Rush would terminate after the command finishes;' +
+        ' adding this parameter will instead watch the file system for changes to the selected projects' +
+        ' (or all projects if no selection was specified).' +
```
Comment (octogonz) [inline apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts:401]: `rush build --watch` without any selectors should build all projects, but it seems to select nothing

File: apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts [hunk 4]
```diff
@@ -329,6 +392,18 @@ export class BulkScriptAction extends BaseScriptAction {
...
+        ' When changes are detected, will re-execute the command on all projects within the selection that are' +
+        ' impacted by the detected changes, then resume waiting for changes.' +
+        ' For details, refer to the website article "Using watch mode".'
+    });
+
```
Comment (octogonz) [inline apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts:406]: In my testing, I found that if the initial invocation of `rush build --watch -T my-app` encounters 1 warning, then the operation will terminate immediately instead of entering a watch loop.

Comment (octogonz) [reply apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts:406]: And if it successfully enters a watch loop, 1 warning seems sufficient to break out of the loop.  This is inconsistent with how `heft build --watch` works.

File: apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts (hunk 3 above)
Comment (dmichon-msft) [reply apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts:415]: Fixed.

File: apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts (hunk 4 above)
Comment (dmichon-msft) [reply apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts:420]: Updated to have any errors from the underlying build get swallowed by the watcher.

Comment (octogonz) [reply apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts:420]: This is still reproing.  Here's my callstack:
```
Error: An error occurred.
    at BulkScriptAction._runOnce (C:\Git\rushstack\apps\rush-lib\lib\cli\scriptActions\BulkScriptAction.js:382:19)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async BulkScriptAction.runAsync (C:\Git\rushstack\apps\rush-lib\lib\cli\scriptActions\BulkScriptAction.js:151:13)
    at async RushCommandLineParser.onExecute (C:\Git\rushstack\libraries\ts-command-line\lib\providers\CommandLineParser.js:199:13)
    at async RushCommandLineParser._wrapOnExecuteAsync (C:\Git\rushstack\apps\rush-lib\lib\cli\RushCommandLineParser.js:140:9)
    at async RushCommandLineParser.onExecute (C:\Git\rushstack\apps\rush-lib\lib\cli\RushCommandLineParser.js:122:13)
    at async RushCommandLineParser.execute (C:\Git\rushstack\libraries\ts-command-line\lib\providers\CommandLineParser.js:100:13)
```

It's going through [this call](https://github.com/microsoft/rushstack/blob/5207e98233858781c7ca1c3d9b4398c6008e370f/apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts#L207) whereas the `try`/`catch` block you added is in `runWatch()`.

Comment (dmichon-msft) [reply apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts:420]: This calls stack isn't in the code path for `--watch`.

**Review by octogonz — APPROVED** [review]

**Comment by octogonz** [comment]
I updated the CLI docs, and fixed some wording in the change log that was referring to internal implementation details. If you're cool with my changes, let's merge this. 👍

Thanks for implementing this -- I can't wait to use it!

File: apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts [hunk 5]
```diff
@@ -169,44 +185,105 @@ export class BulkScriptAction extends BaseScriptAction {
...
-      this._doAfterTask(stopwatch, true);
-    } catch (error) {
-      stopwatch.stop();
+  /**
+   * Runs the command in watch mode. Fundamentally is a simple loop:
+   * 1) Wait for a change to one or more projects in the selection (skipped initially)
+   * 2) Invoke the command on the changed projects, and, if applicable, downstream projects
+   * 3) Goto (1)
+   */
+  protected async runWatch(options: IExecuteInternalOptions): Promise<void> {
```
Comment (iclanton) [inline apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts:217]: If this is `protected`, you need to export `IExecuteInternalOptions` so child classes can access it.

File: apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts [hunk 6]
```diff
@@ -329,6 +406,18 @@ export class BulkScriptAction extends BaseScriptAction {
         ' For details, refer to the website article "Selecting subsets of projects".'
     });
 
+    this._watchParameter = this.defineFlagParameter({
```
Comment (iclanton) [inline apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts:409]: I think we should put this behind an experiment.

File: apps/rush-lib/src/logic/ProjectWatcher.ts [hunk 7]
```diff
@@ -0,0 +1,197 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { FSWatcher } from 'chokidar';
```
Comment (iclanton) [inline apps/rush-lib/src/logic/ProjectWatcher.ts:4]: Use `Import.lazy` to load this.

File: apps/rush-lib/src/logic/ProjectWatcher.ts [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+import { PackageChangeAnalyzer } from './PackageChangeAnalyzer';
+import { RushConfiguration } from '../api/RushConfiguration';
+import { RushConfigurationProject } from '../api/RushConfigurationProject';
+import { Path } from '@rushstack/node-core-library';
```
Comment (iclanton) [inline apps/rush-lib/src/logic/ProjectWatcher.ts:8]:
```suggestion
import { FSWatcher } from 'chokidar';
import { Path } from '@rushstack/node-core-library';

import { PackageChangeAnalyzer } from './PackageChangeAnalyzer';
import { RushConfiguration } from '../api/RushConfiguration';
import { RushConfigurationProject } from '../api/RushConfigurationProject';
```

File: apps/rush-lib/src/logic/ProjectWatcher.ts [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+
+export interface IProjectWatcherOptions {
+  debounceMilliseconds?: number;
+  rushConfiguration: RushConfiguration;
+  selection: ReadonlySet<RushConfigurationProject>;
```
Comment (iclanton) [inline apps/rush-lib/src/logic/ProjectWatcher.ts:13]:
```suggestion
  projectsToMonitor: ReadonlySet<RushConfigurationProject>;
```

File: apps/rush-lib/src/logic/ProjectWatcher.ts [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+            timeout = setTimeout(resolveIfChanged, this._debounceMilliseconds);
+          } catch (err) {
+            terminated = true;
+            reject(err);
+          }
+        });
+      }
+    );
+
+    await watcher.close();
```
Comment (iclanton) [inline apps/rush-lib/src/logic/ProjectWatcher.ts:131]: What is the performance impact of repeatedly opening and closing the watcher?

File: common/changes/@microsoft/rush/watch_2021-01-29-00-52.json [hunk 11]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/rush",
+      "comment": "Add a new parameter \"--watch\" that interactively watches for filesystem changes and rebuilds the affected Rush projects; this feature can also be used with custom bulk commands (GitHub #2458, #1122)",
```
Comment (iclanton) [inline common/changes/@microsoft/rush/watch_2021-01-29-00-52.json:5]: The watching isn't really "interactive," is it? We don't provide an interface for a user to interact with anything in Rush.

File: apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts (hunk 6 above)
Comment (dmichon-msft) [reply apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts:409]: If this is behind an experiment, should we make `runWatch` private?

File: apps/rush-lib/src/logic/ProjectWatcher.ts (hunk 7 above)
Comment (dmichon-msft) [reply apps/rush-lib/src/logic/ProjectWatcher.ts:4]: Using `await import()` to load ProjectWatcher from `BulkScriptAction`. Can also do this for a little extra.

File: apps/rush-lib/package.json [hunk 12]
```diff
@@ -30,6 +30,7 @@
     "@rushstack/ts-command-line": "workspace:*",
     "@yarnpkg/lockfile": "~1.0.2",
     "builtin-modules": "~3.1.0",
+    "chokidar": "~3.4.0",
```
[the remark this answers is not in the document]
Comment (dmichon-msft) [reply apps/rush-lib/package.json:33]: 8 direct dependencies, I don't have a full delta on dependency count

File: apps/rush-lib/src/logic/ProjectWatcher.ts (hunk 10 above)
Comment (dmichon-msft) [reply apps/rush-lib/src/logic/ProjectWatcher.ts:131]: As near as I have been able to tell, not much, but then, we have a deliberate 1 second latency before kicking the build since it may take a non-negligible amount of time.

File: apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts (hunk 5 above)
Comment (dmichon-msft) [reply apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts]: Switched to private for now.

File: apps/rush-lib/src/logic/ProjectWatcher.ts (hunk 9 above)
Comment (dmichon-msft) [reply apps/rush-lib/src/logic/ProjectWatcher.ts]: went with `projectsToWatch` for consistency

File: common/changes/@microsoft/rush/watch_2021-01-29-00-52.json (hunk 11 above)
Comment (dmichon-msft) [reply common/changes/@microsoft/rush/watch_2021-01-29-00-52.json]: Removed the word "interactively"

File: apps/rush-lib/package.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (dmichon-msft) [reply apps/rush-lib/package.json:33]: We already import this version in heft, but rush has its own footprint.

**Comment by wbern** [comment]
Awesome work!!
